### PR TITLE
fix: Ticket Sales: Sold out tickets are not bookable

### LINKED ISSUE #6269: Ticket Sales: Confusing Sold Out Message Instead of Showing "Sold Out"
There is a sold out message when people try to buy a ticket that is sold it. Expected is, that the ticket would not be "bookable" instead there should be a message "Sold Out" in the area drop-down of the ticket number.

For the sold out message itself, if it is displayed anywhere it should show "Ticket [ticket name] sold out." instead of ticket 1865 sold out.

Screenshot from 2021-01-07 10-16-42

### PR DESCRIPTION
**Pull request by MeghalBisht**
Fixes #6269

#### Short description of what this resolves:
Sold out tickets can no longer be booked.

Screenshot -
Capturesw

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by MeghalBisht** [comment]
Few more ways to style the Sold Out text. 
Captureww1
Captureww2
Capturewww

**Comment by iamareebjamal** [comment]
If non-organizers are able to see order statistics, it is an issue, please open a bug for this in server

For showing sold out, there needs to be a new route which returns which ticket is available and which is sold out

**Comment by MeghalBisht** [comment]
By order stats do you mean ticket stats?

File: app/templates/components/public/ticket-list.hbs [hunk 1]
```diff
@@ -80,25 +80,29 @@
...
-                  <div class="default text">{{if (eq this.tickets.length 1) (max ticket.minOrder 1) 0}}</div>
-                  <div class="menu">
-                    <div class="item" data-value="{{map-value mapper 0}}">0</div>
-                    {{#each (range (max ticket.minOrder 1) (add ticket.maxOrder 1)) as |count|}}
-                      <div class="item" data-value="{{map-value mapper count}}">{{count}}</div>
-                    {{/each}}
-                  </div>
-                </UiDropdown>
-              </div>
+              {{#if (sub ticket.quantity (add ticket.orderStatistics.tickets.placed ticket.orderStatistics.tickets.completed))}}
```
Comment (iamareebjamal) [inline app/templates/components/public/ticket-list.hbs:83]: .

Comment (MeghalBisht) [reply app/templates/components/public/ticket-list.hbs:83]: Ok understood, waiting for @SARTHAK1SINGH to follow up else I'll complete it

**Comment by MeghalBisht** [comment]
> If non-organizers are able to see order statistics, it is an issue, please open a bug for this in server
Negative

> For showing sold out, there needs to be a new route which returns which ticket is available and which is sold out
I guess, this needs server modification as well ?

**Comment by iamareebjamal** [comment]
Yes

**Comment by MeghalBisht** [comment]
Working on the server part, will reopen this after that

**Comment by MeghalBisht** [comment]
@iamareebjamal  For the server part, if I add a sold-out column for each ticket will that be a good solution? If not please provide an overview of the approach

**Comment by MeghalBisht** [comment]
As what I think is if I add a column then I need to update its state at multiple places which does not seem to be an optimal solution

**Comment by iamareebjamal** [comment]
/events/ID/tickets/availability route should be created. Which will return each ticket's remaining count

**Comment by MeghalBisht** [comment]
Okay, it should return a boolean as if I return the count then order stats are provided to the purchaser, right?

**Comment by iamareebjamal** [comment]
I thought of boolean but then what if only 4 tickets are left and the user orders 5?

Complete order stats should still not be shown

**Comment by MeghalBisht** [comment]
@iamareebjamal I am sure this is not a good solution, but it works! Please suggest a better way

**Comment by iamareebjamal** [comment]
Move this to a function and call it in init, don't await. Also, set maxQuantity of tickets to min(maxQuantity, availability). and add ticket.remaining, not ticket.available

**Comment by iamareebjamal** [comment]
Also, as the data is discarded after use, there is no use of setting it in the component using this.set

**Comment by MeghalBisht** [comment]
Okay and what about the n² time complexity?

**Comment by MeghalBisht** [comment]
> Also, as the data is discarded after use, there is no use of setting it in the component using this.set

But then how will I get it in the template?

**Comment by iamareebjamal** [comment]
Where are you using it in the template?

File: app/templates/components/public/ticket-list.hbs [hunk 2]
```diff
@@ -80,25 +80,29 @@
...
-                  <div class="default text">{{if (eq this.tickets.length 1) (max ticket.minOrder 1) 0}}</div>
-                  <div class="menu">
-                    <div class="item" data-value="{{map-value mapper 0}}">0</div>
-                    {{#each (range (max ticket.minOrder 1) (add ticket.maxOrder 1)) as |count|}}
-                      <div class="item" data-value="{{map-value mapper count}}">{{count}}</div>
-                    {{/each}}
-                  </div>
-                </UiDropdown>
-              </div>
+              {{#if (lte ticket.available 0)}}
```
Comment (MeghalBisht) [inline app/templates/components/public/ticket-list.hbs:83]: for checking

**Comment by iamareebjamal** [comment]
> Okay and what about the n² time complexity?

Here, n is not 1000000, not even 100.

Besides, doing even 2^n operations on even n=1000 is faster than making an API call to the server which takes 700ms, and is done frequently, so clearly, this is not bottleneck.

Besides, complexity matters when there is a better solution, nested looping over 2 arrays is the only solution to join data, even databases do it, although on more efficient data structures

Other way O(n) will be to loop over availability array and use this.store.peekRecord('ticket', id) to get the ticket ID, but I am not sure if it will return the same object or a different one and maybe template won't reactively change in that case.

Your choice

File: app/templates/components/public/ticket-list.hbs (hunk 2 above)
Comment (iamareebjamal) [reply app/templates/components/public/ticket-list.hbs:83]: That's not `ticketAvailabilities` which you are setting after awaiting

**Comment by MeghalBisht** [comment]
> Other way O(n) will be to loop over availability array and use this.store.peekRecord('ticket', id) to get the ticket ID, but I am not sure if it will return the same object or a different one and maybe template won't reactively change in that case.

I tried this, but when I reload the page multiple times, sometimes it shows SOLD OUT and sometimes it simply shows every ticket available event when they are SOLD OUT. So going with the nested loop.

File: app/components/public/ticket-list.js [hunk 3]
```diff
@@ -14,6 +14,21 @@ export default Component.extend(FormMixin, {
   orderAmount    : null,
   amountOverride : null,
 
+  init() {
```
Comment (iamareebjamal) [inline app/components/public/ticket-list.js:17]: Wasn't there already a didInsertElement here?

Comment (MeghalBisht) [reply app/components/public/ticket-list.js:17]: Yes, its still there, we can use both init and didInsertElement right?

Comment (iamareebjamal) [reply app/components/public/ticket-list.js:17]: Better to move to already existing method

Comment (MeghalBisht) [reply app/components/public/ticket-list.js:17]: you mean to shift code back to didInsertElement? 😕

Comment (iamareebjamal) [reply app/components/public/ticket-list.js]: Yes, the problem was with await. My comment said move the code to another function and call without await.

Comment (MeghalBisht) [reply app/components/public/ticket-list.js]: I am confused now, is the current commit okay?

Comment (iamareebjamal) [reply app/components/public/ticket-list.js]: I have fixed it

Comment (MeghalBisht) [reply app/components/public/ticket-list.js]: Okay now I understood, was confused with your above comments.
Thanks 👍🏻

**Comment by MeghalBisht** [comment]
Please don't merge

**Comment by MeghalBisht** [comment]
I am getting this error after the last commit
zyx

**Comment by iamareebjamal** [comment]
Which event?

**Comment by MeghalBisht** [comment]
I tested with 2 events I created

**Comment by iamareebjamal** [comment]
Fixed, thank you :+1:

**Comment by mariobehling** [comment]
Thanks!

**Pull request merged**